File: src/main/java/com/refinedmods/refinedstorage/inventory/fluid/FilterFluidInventory.java
package com.refinedmods.refinedstorage.inventory.fluid;

import com.refinedmods.refinedstorage.item.FilterItem;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.util.NonNullList;
import net.minecraftforge.fluids.FluidStack;

public class FilterFluidInventory extends FluidInventory {
    public FilterFluidInventory(ItemStack stack) {
        super(27, Integer.MAX_VALUE);

        this.addListener((handler, slot, reading) -> {
            if (!stack.hasTag()) {
                stack.setTag(new CompoundNBT());
            }

            stack.getTag().put(FilterItem.NBT_FLUID_FILTERS, writeToNbt());
        });

        if (stack.hasTag() && stack.getTag().contains(FilterItem.NBT_FLUID_FILTERS)) {
            readFromNbt(stack.getTag().getCompound(FilterItem.NBT_FLUID_FILTERS));
        }
    }

    public NonNullList<FluidStack> getFilteredFluids() {
        NonNullList<FluidStack> list = NonNullList.create();

        for (FluidStack fluid : this.getFluids()) {
            if (fluid != null) {
                list.add(fluid);
            }
        }

        return list;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/inventory/fluid/FilterFluidInventory.kt
package com.refinedmods.refinedstorage.inventory.fluid

import com.refinedmods.refinedstorage.item.FilterItem
import net.minecraft.item.ItemStack
import net.minecraft.nbt.CompoundTag
import net.minecraft.util.NonNullList
import net.minecraftforge.fluids.FluidInstance

class FilterFluidInventory(stack: ItemStack) : FluidInventory(27, Int.MAX_VALUE) {
    fun getFilteredFluids(): NonNullList<FluidInstance> {
        val list: NonNullList<FluidInstance> = NonNullList.create()
        for (fluid in fluids) {
            if (fluid != null) {
                list.add(fluid)
            }
        }
        return list
    }

    init {
        addListener { handler: FluidInventory?, slot: Int, reading: Boolean ->
            if (!stack.hasTag()) {
                stack.tag = CompoundTag()
            }
            stack.tag!!.put(FilterItem.NBT_FLUID_FILTERS, writeToNbt())
        }
        if (stack.hasTag() && stack.tag!!.contains(FilterItem.NBT_FLUID_FILTERS)) {
            readFromNbt(stack.tag!!.getCompound(FilterItem.NBT_FLUID_FILTERS))
        }
    }
}